Update users role

dpkingston opened the pull request:
Unfortunately there are two unrelated changes in this pull request.
9556c99 - this is a trivial change to make the Unix user management not try to install users on anycast instances.
The other two (c9f2057 and 79ac820) should have been in a separate branch and represent the initial support for testing of the routeros_common role.  This is a BIG addition to the code.  It has been tested and is in a working state, but can be optimized and refactored a bit to eliminate some code duplications (notes in the code).  That will com in subsequent CLs.

dpkingston committed:
Don't try to update users on anycast "hosts".
There is no guarantee what host you will wind up connecting to if
multiple hosts provide the service.  Use the hosts specific identity
and address instead.

dpkingston committed:
Commit routeros testing support in a working state.
There is certainly some fine tuning to be done on this but the
ansible + molecule + vagrant support is wired up and working and
successfully processes testing against ROS6 and ROS7.

dpkingston committed:
Additional files needed for initial routeros testing support.

gerner commented on psdr.yml:
```diff
@@ -11,7 +11,7 @@
 - name: Linux User Management
   # https://docs.ansible.com/ansible/latest/inventory_guide/intro_patterns.html#pattern-processing-order
-  hosts: os_linux:&owner_HamWAN:!type_container
+  hosts: os_linux:&owner_HamWAN:!type_container:!type_anycast
```
can you expand on what's going on here?
this is defining which hosts this... playbook (?) applies to? So these are the hosts that will get this "users" role, is that correct?
why does it make sense to exclude type_container (as was already in place)
why does it make sense to exclude type_anycast?
what's the `&` notation here mean?

(just in the context of this PR, no need for a code comment unless you're into that)

dpkingston replied:
The hosts: entry is determining which hosts from the inventory the playbook will be applied to.  The change excludes hosts of type anycast.  This is a reference to the type field in the HamWAN host portal data (host and network database). Example: dns.hamwan.net which is listed as a HamWAN owned Linux host, but this is in fact hosted on srv1.westin.hamwan.net - the real hostname of one of the anycast servers for this address.  Anycast addresses are always present on some real host and that is the one that we should reference for updates.  We omit containers already for other similar reasons.

The HTTP reference in the snippet explains the syntax for host patterns. In this case, in another language this would be written (os is linux and owner is HamWAN) and (type not container and type not anycast)

bkus replied:
> can you expand on what's going on here? this is defining which hosts this... playbook (?) applies to? So these are the hosts that will get this "users" role, is that correct? why does it make sense to exclude type_container (as was already in place) why does it make sense to exclude type_anycast? what's the `&` notation here mean?
> 
> (just in the context of this PR, no need for a code comment unless you're into that)

Ansible docs explain the syntax:

https://docs.ansible.com/ansible/latest/inventory_guide/intro_patterns.html

gerner commented on README.md:
```diff
@@ -87,15 +90,19 @@ cd infrastructure_configs
+
+## See Also
+
+molecule/README.md
```
where is this readme? I can't seem to find it.

also, thanks so much for updating docs :)

dpkingston replied:
Updated this in a new commit.

gerner commented on inventories/test/routeros_hosts.py:
```diff
@@ -0,0 +1,93 @@
+
+if __name__ == "__main__":
+    identities = LocalIdentityFile("/home/dpk/.vagrant.d/insecure_private_key")
+#    identities = VagrantIdentityFiles()
```
do you need this or can you cut it?

dpkingston replied:
I am going to leave it in case its useful in another context.  There is a comment above about why its not being used right now.  Its also present in the users inventory version.

gerner commented on inventories/test/routeros_hosts.py:
```diff
@@ -0,0 +1,93 @@
+
+
+if __name__ == "__main__":
+    identities = LocalIdentityFile("/home/dpk/.vagrant.d/insecure_private_key")
```
this looks specific to your setup, is this how you mean to do this? perhaps this is the line that belongs commented out with a note that this is an example for other people to adapt for their environment while line 67 gets uncommented?

dpkingston replied:
This is sort of a glue record for ansible and vagrant. Let me update this to use HOME environment variable instead to make it more portable.

gerner commented on inventories/test/routeros_hosts.py:
```diff
@@ -0,0 +1,93 @@
+#!/usr/bin/python3
```
a docstring right after this explaining why this exists and what it's for would help future developers (and me)

dpkingston replied:
Done.

gerner commented on inventories/test/routeros_hosts.py:
```diff
@@ -0,0 +1,93 @@
+if __name__ == "__main__":
+    identities = LocalIdentityFile("/home/dpk/.vagrant.d/insecure_private_key")
+#    identities = VagrantIdentityFiles()
+    hosts = {
```
maybe a comment would be useful that says these configs needs need to match the vagrant file config (which is what I think is going on here)

dpkingston replied:
Done.

gerner commented on roles/routeros_common/molecule/default/Vagrantfile:
```diff
@@ -40,11 +61,13 @@ Vagrant.configure("2") do |config|
+      node.vm.provision "routeros_file", name: "Upload initial common configuration", source: "provision/common.rsc", destination: "common.rsc"
+      node.vm.provision "routeros_command", name: "Exec custom script", command: "/import common.rsc", check_script_error: true
+      # Upload and execute a test specific script file
+      node.vm.provision "routeros_file", name: "Upload initial host configuration", source: "provision/#{vm["name"].sub(/ros.\./, '')}.rsc", destination: "host.rsc"
```
matter of opinion, so feel free to disregard: the regex sub here is a little subtle. an explicit mapping from vm to provisioning file would be more explicit and easier to read. I probably spent several minutes confused about how ros6.incorrect gets set up differently than ros6.correct, etc.

as written you're kind of packing a lot of implicit logic into this magical vm name instead of treating that as an opaque string. that always strikes me as a little suspicious.

dpkingston replied:
I think I have a better solution.  Have a look.

gerner commented on roles/routeros_common/molecule/default/converge.yml:
```diff
@@ -3,6 +3,15 @@
+        name: routeros_common
+      vars:
+        test_flaky_network: true
+        # TODO(dpk) - The following should be coming from the inventory, but aren't. Why?
```
is it because the inventory comes from `routeros_hosts.py`? that script is not setting ansible_port and ansible_user, although ansible_ssh_private_key_file does get set there I think.

...upon closer inspection I guess I'm wrong. But I'm not really an ansible expert. the interaction/distinction between `common_vars` and `host_vars` in that script might be an issue...

dpkingston replied:
This appears to no longer be an issue and this comment and the following 3 lines can be removed.  Inventory is working.

gerner commented on roles/routeros_common/molecule/default/converge.yml:
```diff
@@ -3,6 +3,15 @@
+        # TODO(dpk) - The following should be coming from the inventory, but aren't. Why?
+        ansible_port: 22
+        ansible_user: vagrant+ct132w
+        ansible_ssh_private_key_file: "/home/dpk/.vagrant.d/insecure_private_key"
```
this seems like it'll only work on your local environment unless I'm mistaken.

dpkingston replied:
This is fixed.

gerner commented on roles/routeros_common/molecule/default/create.yml:
```diff
@@ -3,21 +3,40 @@
+      ansible.builtin.set_fact:
+        server: {
+          'changed': true,
+          'results': "{{ (inventory.stdout | from_json)._meta.hostvars }}",
```
is this where you're losing `common_vars` that causes the issue of not having `ansible_user` and `ansible_port` above? (notice you're grabbing just hostvars)

dpkingston replied:
No longer broken.

gerner reviewed (changes requested):
I think I get the broad gist of what's going on here. A quick VC to walk through what you're testing would be helpful so I can better understand what's going on in `correct.rsc`, `incorrect.rsc` and `misisng.rsc`. that conversation would probably be more helpful for you than whatever review I can give you at the moment.

dpkingston committed:
Update reference to README for role routeros_common

dpkingston committed:
Add more description of the testing strategy for routeros_common.

dpkingston committed:
Made program portable to other users by using $HOME reference.

dpkingston committed:
Add comment about need to stay in sync with Vagrant configuration.

dpkingston committed:
This makes the choice of file explicit at the minor cost of some repetition.
It has the advantage of also being more flexible if more segregation of config
is needed.

gerner commented:
LGTM, this is ready for merge. Thanks for your patience.
